Client/Frontend/Sync/BraveCore/Bookmarks/BookmarkNodeExtensions.swift

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Data
+import BraveCore
+import BraveShared
+import CoreData
+import Shared
+
+extension BookmarkNode {
+        
+    // MARK: Internal
+    
+    public var domain: Domain? {
+        if let url = titleUrlNodeUrl {
+            return Domain.getOrCreate(forUrl: url, persistent: true)
+        }
+        return nil
+    }
+    
+    public var parentNode: BookmarkNode? {
+        // Return nil if the parent is the ROOT node
+        // because AddEditBookmarkTableViewController.sortFolders
+        // sorts root folders by having a nil parent.
+        // If that code changes, we should change here to match.
+        guard parent?.guid != BookmarkNode.rootNodeGuid else {
+            return nil
+        }
+        
+        return parent
+    }
+    
+    public var objectID: Int {
+        return Int(nodeId)
+    }
+    
+    public func update(customTitle: String?, url: URL?) {
+        setTitle(customTitle ?? "")
+        self.url = url
+    }
+    
+    public func existsInPersistentStore() -> Bool {
+        return isValid && parent != nil
+    }
+    
+    private struct AssociatedObjectKeys {
+        static var faviconObserver: Int = 0
+    }
+    
+    public var bookmarkFavIconObserver: BookmarkModelListener? {
+        get { objc_getAssociatedObject(self, &AssociatedObjectKeys.faviconObserver) as? BookmarkModelListener }
+        set { objc_setAssociatedObject(self, &AssociatedObjectKeys.faviconObserver, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC) }
+    }
+}
+
+class BraveBookmarkFolder: BookmarkNode {
+    public let indentationLevel: Int
+    
+    private init(_ bookmarkNode: BookmarkNode) {
+        self.indentationLevel = 0
+        super.init()
+    }
+    
+    public init(_ bookmarkFolder: BookmarkFolder) {
+        self.indentationLevel = bookmarkFolder.indentationLevel
+        super.init()
+    }
+}
```

Comment:
Is this class necessary? It looks like its just a shortcut for 0 indentation level for a BookmarkNode which seems like its probably only used in 1 place

Reply:
Yeah this looks like necessary since we are sorting our the folder according to class type. Since we are using `BookmarkNode` extension I believe it will be better to subclass for folders instead of adding stored property for checking if it is bookmark with a property like `isFolder` and also storing the indention level.

Reply:
Discussed, we will keep this class but look for a way to remove it in the future with changes to base API in brave-core side